apollo-runtime/src/main/java/com/apollographql/apollo/internal/cache/normalized/RealApolloStore.java

```diff
@@ -225,23 +225,63 @@ public RealApolloStore(@Nonnull NormalizedCache normalizedCache, @Nonnull CacheK
     });
   }
 
-  @Override public <D extends Operation.Data, T, V extends Operation.Variables> void write(
+  @Override @Nonnull public <D extends Operation.Data, T, V extends Operation.Variables> Set<String> write(
       @Nonnull final Operation<D, T, V> operation, @Nonnull final D operationData) {
     checkNotNull(operation, "operation == null");
     checkNotNull(operationData, "operationData == null");
-    writeTransaction(new Transaction<WriteableStore, Boolean>() {
-      @Override public Boolean execute(WriteableStore cache) {
-        CacheResponseWriter cacheResponseWriter = new CacheResponseWriter(operation, customTypeAdapters);
+    return writeTransaction(new Transaction<WriteableStore, Set<String>>() {
+      @Override public Set<String> execute(WriteableStore cache) {
+        CacheResponseWriter cacheResponseWriter = new CacheResponseWriter(operation.variables(), customTypeAdapters);
         try {
           operationData.marshaller().marshal(cacheResponseWriter);
-          Collection<Record> records = cacheResponseWriter.normalize(networkResponseNormalizer());
-          merge(records, CacheHeaders.NONE);
+          ResponseNormalizer<Map<String, Object>> responseNormalizer = networkResponseNormalizer();
+          responseNormalizer.willResolveRootQuery(operation);
+          Collection<Record> records = cacheResponseWriter.normalize(responseNormalizer);
+          return merge(records, CacheHeaders.NONE);
+        } catch (IOException e) {
+          logger.e(e, "Failed to write operation data to the store");
+          throw new RuntimeException(e);
```

Comment:
Actually @digitalbuddha there is next PR that is coming to fix IOException throws mess.
As we moved away from streaming API to read json data there is no need to have readers / writers API to throw IOException as we don't use any IO.